### PR TITLE
feat(replay): Add indicator when the replay is fast-forwarding idle moments

### PR DESCRIPTION
**Pull request by ryan953**
The "Skip to events" toggle seemed like a translation issue for me. The rrweb player isn't really skipping or jumping ahead when the video is idle, instead it's increasing the playback speed until there's something to show. Fast-forward

I thought that this is something worth calling out on the UI, because the fast-forward speed that rrweb picks is variable, I've seen 4x and also 29x speeds in examples.

This replaces the BooleanField with a togglable Button (could use a better icon). 
I don't think it's the best design, but chose it because
a) No label: "Fast-forward through idle moments" seemed like a long label to display at all times
b) I brought the icon into the Player frame. When the fast-forwarding happens you see the new speed, and the icon in the player matches the Button, because they're related.

Screen Shot 2022-04-14 at 2 35 41 PM

Test case was https://dev.getsentry.net:7999/organizations/sentry-emerging-tech/replays/sourcemapsio-replays:e0d1ef068d4b4aeca01d81b26ab2ad1f/

### CODE REVIEW

File: static/app/components/replays/replayContext.tsx [hunk 1]
```diff
@@ -61,13 +63,21 @@ export function Provider({children, events}: React.PropsWithChildren<Props>) {
...
+  const [fastForwardSpeed, setFFSpeed] = useState(0);
 
   const forceDimensions = dimension => {
     setDimensions(dimension as Dimensions);
   };
   const setPlayingFalse = () => {
     setIsPlaying(false);
   };
+  const onFastForwardStart = (e: unknown) => {
+    // @ts-expect-error: rrweb has poor types for event handlers
```
Comment (evanpurkhiser) [inline static/app/components/replays/replayContext.tsx:75]: What is the error exactly?

File: static/app/components/replays/replayPlayer.tsx [hunk 2]
```diff
@@ -76,10 +82,30 @@ function BasePlayerRoot({
...
+const FastForwardBadge = styled('div')`
+  position: absolute;
+  bottom: 0;
+  left: 0;
+  background: ${p => p.theme.purple100};
+  opacity: 0.75;
+  color: ${p => p.theme.textColor};
+  padding: ${space(0.5)};
+  border-bottom-left-radius: ${p => p.theme.borderRadius};
+  border-top-right-radius: ${p => p.theme.borderRadius};
```
Comment (evanpurkhiser) [inline static/app/components/replays/replayPlayer.tsx:106]: Seems like there was an alignment issue in the arrow. Can we use something like

```css
display: grid;
grid-template-columns: max-content max-content;
gap: ${space(0.5)};
align-items: center;
```

This get's things nice and aligned.

Comment (ryan953) [reply static/app/components/replays/replayPlayer.tsx:106]: yes much better! there were some shenanigans with the <Tooltip> too, but here's the change:

**Before:**
before


**After:**
after

Comment (evanpurkhiser) [reply static/app/components/replays/replayPlayer.tsx:106]: That's great!. Design nit, but could maybe use some padding on the right and left too

Comment (ryan953) [reply static/app/components/replays/replayPlayer.tsx]: yeah, got tighter in there with the grid.

roomy now:
Screen Shot 2022-04-15 at 2 50 56 PM

@Jesse-Box hasn't seen this yet. But we'll be back either with a proper icon, or to do some other kind of ui indicator treatment. `@posthog/react-rrweb-player` does an overlay over the whole video for example.

File: static/app/components/replays/replayContext.tsx [hunk 3]
```diff
@@ -33,6 +34,7 @@ const ReplayPlayerContext = React.createContext<ReplayPlayerContextProps>({
   dimensions: {height: 0, width: 0},
   duration: undefined,
   events: [],
+  fastForwardSpeed: 0,
```
Comment (billyvg) [inline static/app/components/replays/replayContext.tsx:37]: Is this different from `speed`? Can we add a comment noting the difference between the two?

File: static/app/components/replays/replayController.tsx [hunk 4]
```diff
@@ -92,15 +91,15 @@ const ReplayControls = ({
...
-          inline={false}
-          stacked
-          hideControlState
-          value={isSkippingInactive}
+        {/* TODO(replay): Need a better icon for the FastForward toggle */}
+        <Button
+          data-test-id="replay-fast-forward"
+          size="xsmall"
+          title={t('Fast Forward idle moments')}
+          aria-label={t('Fast Forward idle moments')}
```
Comment (billyvg) [inline static/app/components/replays/replayController.tsx:99]: Maybe this?

```suggestion
          title={t('Fast-forward idle moments')}
          aria-label={t('Fast-forward idle moments')}
```

Comment (billyvg) [reply static/app/components/replays/replayController.tsx:99]: or 

```suggestion
          title={t('Fast forward idle moments')}
          aria-label={t('Fast forward idle moments')}
```

File: static/app/components/replays/replayContext.tsx (hunk 3 above)
Comment (ryan953) [reply static/app/components/replays/replayContext.tsx:98]: good idea, I added docs to all the values that the context returns.

This one specifically is only set if the video is fast-forwarding right now, which the other speed is the requested playback.

File: static/app/components/replays/replayContext.tsx (hunk 1 above)
Comment (ryan953) [reply static/app/components/replays/replayContext.tsx]: The error is that all the event handlers are typed as `(e: unknown) => void` so there isn't any inference to match teh event type with its payload. 

I ended up writing the handler this way: 
```
const onFastForwardStart = e => {
  setFFSpeed((e as {speed: number}).speed);
};

inst.on(ReplayerEvents.SkipStart, onFastForwardStart);
```

Comment (evanpurkhiser) [reply static/app/components/replays/replayContext.tsx]: If you type it like this is it unhappy?

```tsx
const onFastForwardStart = (e: {speed: number}) => {
  setFFSpeed(e.speed);
};
```

Comment (ryan953) [reply static/app/components/replays/replayContext.tsx]: yes it is unhappy :(

The error moves down to the `.on()` call:
Screen Shot 2022-04-18 at 1 53 59 PM

And the `Handler` type that it mentions is the issue:
Screen Shot 2022-04-18 at 1 53 33 PM

Comment (evanpurkhiser) [reply static/app/components/replays/replayContext.tsx]: lol so annoying. IMO I would keep the ts-expect-error in that case

``tsx
// @ts-expect-error: rrweb types event handlers with `unknown` parameters
inst.on(ReplayerEvents.SkipStart, onFastForwardStart);
```

**Review by billyvg — APPROVED** [review]
Good call on adding the comments

File: static/app/components/replays/replayContext.tsx (hunk 1 above)
Comment (ryan953) [reply static/app/components/replays/replayContext.tsx]: updated

**Pull request merged**